src/Transform/DisconnectKrnlDimFromAlloc.cpp

```diff
@@ -0,0 +1,104 @@
+//===-------- DisconnectKrnlDimFromAlloc.cpp ------------------------------===//
+//
+// Copyright 2019-2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This pass enables the lowering of the krnl.dim operation to a series of
+// instruction which do not depend on the alloc of the MemRef whose dim is
+// being taken. The krnl.dim operation works in the presence of MemRefs
+// which contain affine maps by ignoring the map if present.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/Krnl/KrnlOps.hpp"
+#include "src/Pass/Passes.hpp"
+
+using namespace mlir;
+
+namespace {
+
+/*!
+ *  RewritePattern that replaces:
+ *    %0 = alloc(%d) : memref<?x10x<type>, #map>
+ *    %1 = krnl.dim(%0, 0) : memref<?x10x<type>>
+ *    %2 = krnl.dim(%0, 1) : memref<?x10x<type>>
+ *    %3 = add %1, %2
+ *  with:
+ *    %0 = alloc(%d) : memref<?x10x<type>, #map>
+ *    %2 = constant 10 : index
+ *    %3 = add %d, %2
+ */
+
+class DisconnectKrnlDimFromAlloc : public OpRewritePattern<KrnlDimOp> {
+public:
+  using OpRewritePattern<KrnlDimOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      KrnlDimOp krnlDimOp, PatternRewriter &rewriter) const override {
+    auto loc = krnlDimOp.getLoc();
+
+    // If index is not constant, return failure.
+    // TODO: support dynamic index case.
```

Comment:
I can't imagine why we would need to support a dynamic index: since we are looking at the memref, looking at the '?', and then adding them to the alloc function as parameter.